File: services/mas-service/llm/agents/config_agent_network.js
/**
 * Zentrale Netzwerkagenten-Konfiguration
 * 
 * Hier werden alle Agents des Netzwerks zentral definiert.
 * Ein Master-Agent koordiniert die Pipeline und verteilt Aufgaben an spezialisierte Worker-Agents.
 * 
 * Vorteile:
 * - Zentrale Konfiguration aller Agents
 * - Einfache Erweiterung um neue Agents
 * - Klare Trennung von Koordination und Ausführung
 * - Wartbare und skalierbare Architektur
 */

import { getSystemInstruction } from './system_instructions.js';

// Standard-Fallback-Modell
const DEFAULT_MODEL = 'mistral:latest';

/**
 * Master-Agent: Koordiniert die gesamte ML-Pipeline
 * Kennt die Pipeline-Schritte und verteilt Aufgaben an Worker-Agents
 */
export const MASTER_AGENT = {
  key: 'MASTER_AGENT',
  name: 'Master-Pipeline-Koordinator',
  model: 'mistral:latest',
  description: 'Koordiniert die ML-Pipeline und verteilt Aufgaben an spezialisierte Agents',
  systemPrompt: getSystemInstruction('MASTER_AGENT'),
  temperature: 0.1,
  maxTokens: 1024,
  retries: 3,
  timeout: 30000,
  category: 'coordination',
  icon: '🎯',
  role: 'master'
};

/**
 * Worker-Agents: Spezialisierte Agents für spezifische Aufgaben
 */
export const WORKER_AGENTS = {
  // Datenanalyse-Agent
  DATA_ANALYZER: {
    key: 'DATA_ANALYZER',
    name: 'Datenanalyse-Agent',
    // model: 'qwen3:4b',
    model: 'llama3.2:latest',
    description: 'Analysiert Datasets und erstellt detaillierte Insights',
    systemPrompt: getSystemInstruction('DATA_ANALYZER'),
    temperature: 0.2,
    maxTokens: 4096,
    retries: 3,
    timeout: 60000,
    category: 'analysis',
    icon: '📊',
    role: 'worker',
    capabilities: ['data_analysis', 'pattern_recognition', 'statistical_analysis'],
    inputRequirements: ['dataset_path', 'data_analysis'],
    outputType: 'data_insights'
  },

  // Hyperparameter-Optimierer
  HYPERPARAMETER_OPTIMIZER: {
    key: 'HYPERPARAMETER_OPTIMIZER',
    name: 'Hyperparameter Optimizer',
    // model: 'qwen3:4b',
    model: 'llama3.2:latest',
    description: 'Suggests optimal hyperparameters and features based on data analysis',
    systemPrompt: getSystemInstruction('HYPERPARAMETER_OPTIMIZER'),
    temperature: 0.1,
    maxTokens: 2048,
    retries: 3,
    timeout: 60000,
    category: 'optimization',
    icon: '⚙️',
    role: 'worker',
    capabilities: ['hyperparameter_tuning', 'algorithm_selection', 'optimization'],
    inputRequirements: ['data_insights', 'algorithm_type'],
    outputType: 'hyperparameter_suggestions'
  },

  // Code-Generator
  CODE_GENERATOR: {
    key: 'CODE_GENERATOR',
    name: 'Python Code Generator',
    // model: 'qwen3:4b',
    model: 'llama3.2:latest',
    description: 'Generates optimized Python code for ML training',
    systemPrompt: getSystemInstruction('CODE_GENERATOR'),
    temperature: 0.0,
    maxTokens: 4096,
    retries: 3,
    timeout: 60000,
    category: 'code',
    icon: '🔧',
    role: 'worker',
    capabilities: ['code_generation', 'ml_implementation', 'python_programming'],
    inputRequirements: ['hyperparameter_suggestions', 'data_insights', 'project_config'],
    outputType: 'python_code'
  },

  // Code-Reviewer
  CODE_REVIEWER: {
    key: 'CODE_REVIEWER',
    name: 'Code Review Agent',
    // model: 'qwen3:4b',
    model: 'llama3.2:latest',
    description: 'Reviews and optimizes generated Python code',
    systemPrompt: getSystemInstruction('CODE_REVIEWER'),
    temperature: 0.2,
    maxTokens: 4096,
    retries: 2,
    timeout: 60000,
    category: 'review',
    icon: '🔍',
    role: 'worker',
    capabilities: ['code_review', 'optimization', 'best_practices'],
    inputRequirements: ['python_code'],
    outputType: 'reviewed_code'
  },

  // Performance-Analyzer
  PERFORMANCE_ANALYZER: {
    key: 'PERFORMANCE_ANALYZER',
    name: 'Performance Analyzer',
    // model: 'qwen3:4b',
    model: 'llama3.2:latest',
    description: 'Analyzes model performance and gives improvement suggestions',
    systemPrompt: getSystemInstruction('PERFORMANCE_ANALYZER'),
    temperature: 0.3,
    maxTokens: 4096,
    retries: 2,
    timeout: 60000,
    category: 'analysis',
    icon: '📈',
    role: 'worker',
    capabilities: ['performance_analysis', 'metric_evaluation', 'improvement_suggestions'],
    inputRequirements: ['model_results', 'performance_metrics'],
    outputType: 'performance_analysis'
  }
};

/**
 * Pipeline-Definition: Definiert die Standard-ML-Pipeline-Schritte
 */
export const PIPELINE_STEPS = [
  {
    step: 1,
    name: 'Data Analysis',
    agent: 'DATA_ANALYZER',
    required: true,
    description: 'Analysiert das Dataset und erstellt Insights'
  },
  {
    step: 2,
    name: 'Hyperparameter Optimization',
    agent: 'HYPERPARAMETER_OPTIMIZER',
    required: true,
    description: 'Suggests optimal hyperparameters',
    dependsOn: ['DATA_ANALYZER']
  },
  {
    step: 3,
    name: 'Code Generation',
    agent: 'CODE_GENERATOR',
    required: true,
    description: 'Generates Python code for ML training',
    dependsOn: ['HYPERPARAMETER_OPTIMIZER']
  },
  // {
  //   step: 4,
  //   name: 'Code-Review',
  //   agent: 'CODE_REVIEWER',
  //   required: false,
  //   description: 'Überprüft und optimiert den generierten Code',
  //   dependsOn: ['CODE_GENERATOR']
  // },
  // {
  //   step: 5,
  //   name: 'Performance-Analyse',
  //   agent: 'PERFORMANCE_ANALYZER',
  //   required: false,
  //   description: 'Analysiert Modell-Performance',
  //   // dependsOn: ['CODE_REVIEWER', 'CODE_GENERATOR']
  //   dependsOn: ['CODE_GENERATOR']
  // }
];

/**
 * Alle verfügbaren Agents (Master + Worker)
 */
export const ALL_AGENTS = {
  [MASTER_AGENT.key]: MASTER_AGENT,
  ...WORKER_AGENTS
};

export function getAgentConfig(agentKey) {
  const config = ALL_AGENTS[agentKey];
  if (!config) {
    console.warn(`⚠️ Agent configuration for '${agentKey}' not found.`);
    return {
      key: agentKey,
      name: `Unknown Agent (${agentKey})`,
      model: DEFAULT_MODEL,
      description: 'Standard agent with default configuration',
      systemPrompt: 'You are a helpful AI assistant.',
      temperature: 0.7,
      maxTokens: 2048,
      retries: 2,
      timeout: 30000,
      role: 'unknown'
    };
  }
  return { ...config };
}

export function getAgentModel(agentKey) {
  const config = getAgentConfig(agentKey);
  return config.model || DEFAULT_MODEL;
}

export function getWorkerAgents() {
  return Object.values(WORKER_AGENTS);
}

export function getAgentsByCategory(category) {
  return Object.values(ALL_AGENTS).filter(agent => agent.category === category);
}

export function getAgentsByCapability(capability) {
  return Object.values(WORKER_AGENTS).filter(agent =>
    agent.capabilities && agent.capabilities.includes(capability)
  );
}

export function isValidAgent(agentKey) {
  return agentKey in ALL_AGENTS;
}

export function getPipelineSteps() {
  return [...PIPELINE_STEPS];
}

export function getNextPipelineStep(completedSteps = []) {
  const completedAgentKeys = completedSteps.map(step => step.agent);

  for (const step of PIPELINE_STEPS) {
    const dependenciesMet = !step.dependsOn ||
      step.dependsOn.every(dep => completedAgentKeys.includes(dep));

    // Prüfe, ob dieser Schritt noch nicht abgeschlossen ist
    const notCompleted = !completedAgentKeys.includes(step.agent);

    if (dependenciesMet && notCompleted) {
      return step;
    }
  }

  return null;
}

export function logAgentCall(agentKey, model, operation) {
  const timestamp = new Date().toISOString();
  console.log(`🤖 [${timestamp}] Agent call: ${agentKey} (${model}) - ${operation}`);
}

export function getAgentStats() {
  const totalAgents = Object.keys(ALL_AGENTS).length;
  const workerAgents = Object.keys(WORKER_AGENTS).length;
  const categories = [...new Set(Object.values(ALL_AGENTS).map(a => a.category))];

  return {
    totalAgents,
    masterAgents: 1,
    workerAgents,
    categories: categories.length,
    pipelineSteps: PIPELINE_STEPS.length
  };
}

File: services/mas-service/core/agents/config_agent_network.py
"""
Zentrale Netzwerkagenten-Konfiguration mit Model-Tier-System
"""

import os

# Einheitliches Modell für alle Agents - vermeidet 200+ Sekunden Modell-Wechsel in Ollama
DEFAULT_MODEL = 'deepseek-r1:1.5b'

# Worker-Agents Konfiguration
WORKER_AGENTS = {
    'DATA_ANALYZER': {
        'key': 'DATA_ANALYZER',
        'name': 'Datenanalyse-Agent',
        'model': 'deepseek-r1:1.5b',  # Einheitliches Modell für alle Agents
        'description': 'Analysiert Datasets und erstellt detaillierte Insights',
        'temperature': 0.2,
        'maxTokens': 4096,
        'retries': 3,
        'timeout': 90000, # 1,5min
        'category': 'analysis',
        'icon': '📊',
        'role': 'worker'
    },
    'DATA_CLEANER': {
        'key': 'DATA_CLEANER',
        'name': 'Data Cleaner',
        'model': 'deepseek-r1:1.5b',  # Einheitliches Modell für alle Agents
        'description': 'Cleans and preprocesses data based on quality analysis',
        'temperature': 0.1,
        'maxTokens': 1024,
        'retries': 3,
        'timeout': 60000,
        'category': 'preprocessing',
        'icon': '🧹',
        'role': 'worker'
    },
    'FEATURE_ENGINEER': {
        'key': 'FEATURE_ENGINEER',
        'name': 'Feature Engineer',
        'model': 'deepseek-r1:1.5b',  # Einheitliches Modell für alle Agents
        'description': 'Generiert neue Features aus vorhandenen Daten',
        'temperature': 0.2,
        'maxTokens': 4096,
        'retries': 3,
        'timeout': 90000, # 1,5min
        'category': 'engineering',
        'icon': '🔧',
        'role': 'worker'
    },
    'HYPERPARAMETER_OPTIMIZER': {
        'key': 'HYPERPARAMETER_OPTIMIZER',
        'name': 'Hyperparameter Optimizer',
        'model': 'deepseek-r1:1.5b',  # Einheitliches Modell für alle Agents
        'description': 'Suggests optimal hyperparameters',
        'temperature': 0.1,
        'maxTokens': 2048,
        'retries': 3,
        'timeout': 60000,
        'category': 'optimization',
        'icon': '⚙️',
        'role': 'worker'
    },
    'CODE_GENERATOR': {
        'key': 'CODE_GENERATOR',
        'name': 'Code Generator',
        'model': 'deepseek-r1:1.5b',   # Kein LLM verwendet (Template-basiert)
        'description': 'Generates Python code from template (no LLM)',
        'temperature': 0.1,
        'maxTokens': 8192,
        'retries': 3,
        'timeout': 100000, # 100s
        'category': 'generation',
        'icon': '💻',
        'role': 'worker'
    },
    'CODE_REVIEWER': {
        'key': 'CODE_REVIEWER',
        'name': 'Code Reviewer',
        'model': 'deepseek-r1:1.5b',  # Einheitliches Modell für alle Agents
        'description': 'Reviews generated code for correctness, security, and best practices',
        'temperature': 0.1,
        'maxTokens': 2048,
        'retries': 3,
        'timeout': 60000,
        'category': 'review',
        'icon': '👨‍💻',
        'role': 'worker'
    },
    'CODE_EXECUTOR': {
        'key': 'CODE_EXECUTOR',
        'name': 'Code Executor',
        'model': 'deepseek-r1:1.5b',
        'description': 'Executes Python code (no LLM)',
        'temperature': 0.0,
        'maxTokens': 0,
        'retries': 3,
        'timeout': 60000,
        'category': 'execution',
        'icon': '▶️',
        'role': 'worker'
    },
    'PERFORMANCE_ANALYZER': {
        'key': 'PERFORMANCE_ANALYZER',
        'name': 'Performance Analyzer',
        'model': 'deepseek-r1:1.5b',  # Einheitliches Modell für alle Agents
        'description': 'Analyzes model performance',
        'temperature': 0.1,
        'maxTokens': 1024,
        'retries': 3,
        'timeout': 60000,
        'category': 'analysis',
        'icon': '📈',
        'role': 'worker'
    },
    'DECISION': {
        'key': 'DECISION',
        'name': 'Decision Agent',
        'model': 'deepseek-r1:1.5b',  # Einheitliches Modell für alle Agents
        'description': 'Decides if result is good enough or loop again',
        'temperature': 0.1,
        'maxTokens': 512,
        'retries': 3,
        'timeout': 60000,
        'category': 'decision',
        'icon': '🎯',
        'role': 'worker'
    }
}

# Alle Agents
ALL_AGENTS = {**WORKER_AGENTS}

# Pipeline-Schritte
PIPELINE_STEPS = [
    {'step': 1, 'name': 'Datenanalyse', 'agent': 'DATA_ANALYZER', 'required': True},
    {'step': 2, 'name': 'Feature Engineering', 'agent': 'FEATURE_ENGINEER', 'required': True},
    {'step': 3, 'name': 'Datenbereinigung', 'agent': 'DATA_CLEANER', 'required': False},
    {'step': 4, 'name': 'Hyperparameter-Optimierung', 'agent': 'HYPERPARAMETER_OPTIMIZER', 'required': True},
    {'step': 5, 'name': 'Code-Generierung', 'agent': 'CODE_GENERATOR', 'required': True},
    {'step': 6, 'name': 'Code-Review', 'agent': 'CODE_REVIEWER', 'required': False},
    {'step': 7, 'name': 'Code-Ausführung', 'agent': 'CODE_EXECUTOR', 'required': True},
    {'step': 8, 'name': 'Performance-Analyse', 'agent': 'PERFORMANCE_ANALYZER', 'required': True},
    {'step': 9, 'name': 'Entscheidung', 'agent': 'DECISION', 'required': True}
]

# Agent-Statistiken (vereinfacht)
_agent_stats = {
    'totalCalls': 0,
    'successfulCalls': 0,
    'failedCalls': 0
}

def get_agent_config(agent_key: str) -> dict:
    """Hole Agent-Konfiguration"""
    return ALL_AGENTS.get(agent_key, {})

def get_agent_model(agent_key: str, use_tier: bool = None) -> str:
    """
    Hole Agent-Modell basierend auf Model-Tier-System
    
    Args:
        agent_key: Agent-Schlüssel
    
    Returns:
        Modell-Name für den Agent
    """
    config = get_agent_config(agent_key)
    
    # Fallback: Verwende konfiguriertes Modell
    return config.get('model', DEFAULT_MODEL)

def is_valid_agent(agent_key: str) -> bool:
    """Prüfe ob Agent existiert"""
    return agent_key in ALL_AGENTS

def log_agent_call(agent_key: str, model: str, purpose: str):
    """Logge Agent-Aufruf"""
    _agent_stats['totalCalls'] += 1
    print(f'🤖 {agent_key} ({model}): {purpose}')

def get_agent_stats() -> dict:
    """Hole Agent-Statistiken"""
    return _agent_stats.copy()

def get_model_tier_info(tier: str = None) -> dict:
    """
    Hole Informationen über Model-Tiers
    
    Args:
        tier: Optional - Spezifischer Tier (HIGH/MEDIUM/LOW)
    
    Returns:
        Tier-Konfiguration oder alle Tiers
    """
    if tier:
        return MODEL_TIERS.get(tier, {})
    return MODEL_TIERS.copy()

def get_agent_tier(agent_key: str) -> str:
    """Hole Model-Tier für einen Agent"""
    config = get_agent_config(agent_key)
    return config.get('modelTier', 'HIGH')

def get_agents_by_tier(tier: str) -> list:
    """Hole alle Agents mit einem bestimmten Model-Tier"""
    return [
        agent_key 
        for agent_key, config in ALL_AGENTS.items() 
        if config.get('modelTier') == tier
    ]

def get_optimization_summary() -> dict:
    """
    Erstelle Zusammenfassung der Model-Tier-Optimierung
    
    Returns:
        Dictionary mit Optimierungs-Statistiken
    """
    total_agents = len(ALL_AGENTS)
    llm_agents = [k for k, v in ALL_AGENTS.items() if v.get('modelTier') is not None]
    
    tier_counts = {
        'HIGH': len(get_agents_by_tier('HIGH')),
        'MEDIUM': len(get_agents_by_tier('MEDIUM')),
        'LOW': len(get_agents_by_tier('LOW'))
    }
    
    optimized_agents = tier_counts['MEDIUM'] + tier_counts['LOW']
    optimization_percentage = (optimized_agents / len(llm_agents) * 100) if llm_agents else 0
    
    return {
        'totalAgents': total_agents,
        'llmAgents': len(llm_agents),
        'tierCounts': tier_counts,
        'optimizedAgents': optimized_agents,
        'optimizationPercentage': round(optimization_percentage, 1),
        'useTiers': USE_MODEL_TIERS,
        'agentsByTier': {
            'HIGH': get_agents_by_tier('HIGH'),
            'MEDIUM': get_agents_by_tier('MEDIUM'),
            'LOW': get_agents_by_tier('LOW')
        }
    }

